### PR TITLE
Glob parser

### LINKED ISSUE #55: the word_cst parser does not recognize globs and character ranges
`*`, `?` and character ranges are always parsed as literals

### PR DESCRIPTION
**Pull request by yurug**
This PR is an attempt to implement pattern matching recognition.

Fixes #55.

### CODE REVIEW

**Comment by yurug** [comment]
Regarding this part of the standard:
```
[
If an open bracket introduces a bracket expression as in XBD RE Bracket Expression, except that the 
<exclamation-mark> character ( '!' ) shall replace the <circumflex> character ( '^' ) in its role in a non-
matching list in the regular expression notation, it shall introduce a pattern bracket expression. A bracket 
expression starting with an unquoted <circumflex> character produces unspecified results. Otherwise, '[' 
shall match the character itself.
```
I did not choose to interpret `^` as `!` even if the semantics of `^` is unspecified by the standard. `bash` and `dash` took a different path, choosing to see `^` as `!`.

@nios @treinen, do you agree with that design choice?

**Comment by treinen** [comment]
I think it s the right decision to follow the POSIX standard. However, it might be worth to add in the future an option to morbig, like "--error-on-unspecified", which makes morbig produce an error 
(or at least a warning) when the syntactic structure of an input script depends on a decision which is
explicitely labeled as unspecified in the standard.

So, OK to merge for now and leave that option as a feature request for the future.

File: src/REBracketExpressionParser.mly [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+}
+;
+
+/* The following rule is not in the grammar of the POSIX standard as
+   explained in the header of this file. */
+meta_char: MINUS {
+  '-'
+}
+| HAT {
+  '^'
```
Comment (Niols) [inline src/REBracketExpressionParser.mly]: Hum, I don't understand, shouldn't that be `'!'`?

File: src/dune [hunk 2]
```diff
@@ -1,8 +1,8 @@
-(ocamllex prelexer)
+(ocamllex prelexer REBracketExpressionLexer)
 
 (menhir
  (flags --explain --inspection --table)
- (modules parser))
+ (modules parser REBracketExpressionParser))
```
Comment (Niols) [inline src/dune]: Btw, do we need all these flags for the parser?

Comment (Niols) [reply src/dune]: I mean, I just kept them as they were in the Makefile before, but I'm not sure about what they do.

File: src/prelexerState.ml [hunk 3]
```diff
@@ -1,3 +1,16 @@
+(**************************************************************************)
+(*  -*- tuareg -*-                                                        *)
+(*                                                                        *)
+(*  Copyright (C) 2017,2018 Yann Régis-Gianas, Nicolas Jeannerod,         *)
```
Comment (Niols) [inline src/prelexerState.ml]: @treinen Maybe we should run headache before releasing? To have 2017-2019?

File: src/CST.ml [hunk 4]
```diff
@@ -329,11 +329,64 @@ and word_component =
...
+and equivalence_class =
+  | EquivalenceClass_OpenEqual_COLLELEMSINGLE_EqualClose of char
+  | EquivalenceClass_OpenEqual_COLLELEMMULTI_EqualClose of string
+
+and character_class =
+  | CharacterClass_OpenColon_CLASSNAME_ColonClose of class_name
+
+and class_name =
+  | ClassName of string (* Keyword in the current LC_CTYPE category. *)
+
```
Comment (Niols) [inline src/CST.ml]: Damn, that's a lot of new types for such a "small" thing (meaning it wasn't that small after all).

File: src/REBracketExpressionLexer.mll [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+| "=]"   { Equal_close }
+| "[."   { Open_dot    }
+| ".]"   { Dot_close   }
+| "[:"   { Open_colon  }
+| ":]"   { Colon_close }
+
+(** The following rule returns a META_CHAR that will be
+    turned into a COLL_ELEM_SINGLE or even a COLL_ELEM_MULTI
+    depending on the parsing context. *)
+| _ as c { COLL_ELEM_SINGLE c }
```
Comment (Niols) [inline src/REBracketExpressionLexer.mll]: So `^` ends up in there, am I right?

File: src/REBracketExpressionParser.mly (hunk 1 above)
Comment (yurug) [reply src/REBracketExpressionParser.mly]: Fixed by bf69544

File: src/dune (hunk 2 above)
Comment (yurug) [reply src/dune]: Yes, we need them. `--inspection` is important to the incremental parser engine. `--explain` is to get `.conflicts` files. `--table` is to use the interpreter back-end.

File: src/REBracketExpressionLexer.mll (hunk 5 above)
Comment (yurug) [reply src/REBracketExpressionLexer.mll]: Yes.

File: src/dune (hunk 2 above)
Comment (Niols) [reply src/dune]: Aren't `.conflict` files meant more for debugging?

**Comment by yurug** [comment]
I have implemented the command line option you suggested @treinen.

File: src/dune (hunk 2 above)
Comment (yurug) [reply src/dune]: Yes, but they are not really expansive to produce. But, OK, this can moved to the debugging mode of the build system, right.

**Comment by treinen** [comment]
On Fri, Feb 22, 2019 at 05:31:07AM -0800, Nicolas Jeannerod wrote:
> Niols commented on this pull request.
> 
> 
> 
> > @@ -1,3 +1,16 @@
> +(**************************************************************************)
> +(*  -*- tuareg -*-                                                        *)
> +(*                                                                        *)
> +(*  Copyright (C) 2017,2018 Yann Régis-Gianas, Nicolas Jeannerod,         *)
> 
> @treinen Maybe we should run headache before releasing? To have 2017-2019?

Done (on the master branch only). -Ralf.
-- 
Ralf Treinen
Institut de Recherche en Informatique Fondamentale
Équipe Preuves, Programmes et Systèmes
Université Paris Diderot, Paris, France.
http://www.irif.fr/~treinen/

**Comment by Niols** [comment]
@yurug I have nothing more to say about this PR. imho, it can be merge (we'd still have to wait for the CI to go green).

**Comment by Niols** [comment]
CI is happy and so am I.

**Comment by yurug** [comment]
Great! Let's merge then once the green lights are here.

**Comment by Niols** [comment]
They are. It's just the GitHub-Travis hook that failed. But if you click on details, everything is fine.

**Pull request merged**